### PR TITLE
ARROW-3564: [C++] Fix dictionary encoding logic for Parquet 2.0

### PR DESCRIPTION
**Pull request by hatemhelal**
Reworked logic for encoder selection in column writer.  Added C++ tests and expect that this should resolve the problem observed with using dictionary encoding to write parquet files using pyarrow described in ARROW-3564.

### CODE REVIEW

**Comment by hatemhelal** [comment]
The CI failures are unrelated to this PR.

**Comment by xhochy** [comment]
@hatemhelal Can you rebase this branch on master to get rid of the CI failures?

File: cpp/src/parquet/column_writer.cc [hunk 1]
```diff
@@ -534,22 +534,16 @@ void ColumnWriter::FlushBufferedDataPages() {
...
-      current_encoder_.reset(new PlainEncoder<Type>(descr_, properties->memory_pool()));
-      break;
-    case Encoding::PLAIN_DICTIONARY:
-    case Encoding::RLE_DICTIONARY:
-      current_encoder_.reset(new DictEncoder<Type>(descr_, properties->memory_pool()));
-      break;
-    default:
-      ParquetException::NYI("Selected encoding is not supported");
+    : ColumnWriter(metadata, std::move(pager), use_dictionary, encoding, properties) {
+  if (use_dictionary) {
```
Comment (xhochy) [inline cpp/src/parquet/column_writer.cc]: Why did you change the old logic from the switch to the extra `use_dictionary` flag?

**Review by xhochy — COMMENTED** [review]
I don't yet understand what really went wrong but that is probably handled by my inline question.

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (hatemhelal) [reply cpp/src/parquet/column_writer.cc]: This wasn't obvious to me either.  For the version 2.0 code path it seems that we need to instantiate `ColumnWriter` with the `RLE_DICTIONARY` encoding but previously it was using the `PLAIN` encoding as defined here in arrow/cpp/src/parquet/properties.h:

```
  inline Encoding::type dictionary_page_encoding() const {
    if (parquet_version_ == ParquetVersion::PARQUET_1_0) {
      return Encoding::PLAIN_DICTIONARY;
    } else {
      return Encoding::PLAIN;
    }
  }
```

To fix this, I replaced the call to `dictionary_page_encoding()` with `dictionary_index_encoding()` in `ColumnWriter::Make'.  It seemed clearer to me to also compute the `use_dictionary` option there too and pass it down the stack.  Let me know if you have any suggestions to make this clearer?

Comment (xhochy) [reply cpp/src/parquet/column_writer.cc]: Thanks for clarifying, that makes a bit more sense to me now.

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Comment by hatemhelal** [comment]
Thanks @xhochy, appreciate the code review!
